libbeat/common/capabilities/capabilities_linux.go

```diff
@@ -0,0 +1,209 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build linux
+
+package capabilities
+
+import (
+	"errors"
+	"math/bits"
+	"strconv"
+	"strings"
+
+	"kernel.org/pub/linux/libs/security/libcap/cap"
+)
+
+var (
+	// Tried to map an invalid capability ID: x < 0 || x >= 64.
+	errInvalidCapability = errors.New("invalid capability")
+
+	// Capabilities are linux only and this is returned on other
+	// systems for all the public functions. There is a generic
+	// errors.ErrUnsupported present in golang 1.21, but we still
+	// support 1.20.
+	ErrUnsupported = errors.New("capabilities are only supported in linux")
+
+	// The mask when all known capabilities are set.
+	allMask = (uint64(1) << uint64(cap.MaxBits())) - 1
+)
+
+// The capability set flag/vector, re-exported from
+// libcap(3). Inherit, Bound & Ambient not exported since we have no
+// use for it yet.
+type Flag = cap.Flag
+
+const (
+	// aka CapEff
+	Effective = cap.Effective
+	// aka CapPrm
+	Permitted = cap.Permitted
+)
+
+// True if sets are equal for the given flag/vector, errors out in
+// case any of the sets is malformed.
+func isEqual(flag Flag, a *cap.Set, b *cap.Set) (bool, error) {
+	d, err := a.Cf(b)
+	if err != nil {
+		return false, err
+	}
+
+	return !d.Has(flag), nil
+}
+
+// Convert the capability ID to a string suitable to be used in
+// ECS.
+// If capabiliy ID X is unknown, but valid (0 <= X < 64), "CAP_X"
+// will be returned instead. Fetches from an internal table built at
+// startup.
+var toECS = makeToECS()
+
+// Make toECS() which creates a map of every possible valid capability
+// ID on startup. Returns errInvalidCapabilty for an invalid ID.
+func makeToECS() func(int) (string, error) {
+	ecsNames := make(map[int]string)
+
+	for i := 0; i < 64; i++ {
+		c := cap.Value(i)
+		if i < int(cap.MaxBits()) {
+			ecsNames[i] = strings.ToUpper(c.String())
+		} else {
+			ecsNames[i] = strings.ToUpper("CAP_" + c.String())
+		}
+	}
+
+	return func(b int) (string, error) {
+		s, ok := ecsNames[b]
+		if !ok {
+			return "", errInvalidCapability
+		}
+		return s, nil
+	}
+}
+
+// True if the set has all the capabilities set for the given
+// flag/vector, see FromUint64 for a CAP_ALL explanation.
+var isAll = makeIsAll()
+
+// Make isAll(), there is no direct way to get a full capability set,
+// so we have to build one. Instead of building it for every call,
+// build it once on startup and don't expose it.
+func makeIsAll() func(Flag, *cap.Set) (bool, error) {
+	all := cap.NewSet()
+	for i := 0; i < int(cap.MaxBits()); i++ {
+		all.SetFlag(cap.Effective, true, cap.Value(i))
+		all.SetFlag(cap.Permitted, true, cap.Value(i))
+		all.SetFlag(cap.Inheritable, true, cap.Value(i))
+	}
+
+	return func(flag Flag, set *cap.Set) (bool, error) {
+		return isEqual(flag, set, all)
+	}
+}
+
+// Like isAll(), but for the empty set, here for symmetry.
+var isEmpty = makeIsEmpty()
```

Comment:
I'd have to `s/emtpy/cap.NewSet()`, which allocates a set per call, I'd be fine with it but since there is already the same scheme with `isAll`, I'd leave it is.